### ISSUE TITLE
options with * are not applied to the root package

### ISSUE THREAD

**Issue by wpalfi** [conan-io/conan#2391]
Example:
I want to install some modular Boost package which requires other Boost packages. I want all of them to be shared. The intuitive command would be
`conan install Boost.Test/1.65.1@user/testing -o Boost.*:shared=True
`
But that installs only the requirements shared, the main package (Boost.Test) is static. To avoid that, I need to do
`conan install Boost.Test/1.65.1@user/testing -o Boost.*:shared=True -o shared=True
`
Is that the intended behaviour? If so, I would at least add a warning in the docs.

Conan 1.0.3, Windows
- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
Yes, this seems to be a small bug. I will try to investigate, if possible to fix for 1.0.4, if not it will have to wait until 1.1.

Thanks very much for reporting.

**Comment by uilianries** [comment]
Hi!

@solvingj and me are talking about this problem now.

During modular Boost build, we have 8 jobs by build (default sort by Conan package tools). However, when the upload starts, only 4 packages are built.

The problem is that we use `boost_*:shared` as named shared option. This pattern could be filled using False or True, but both values results on same package id.

Here are one example: https://travis-ci.org/bincrafters/conan-boost_level8group/jobs/334141358
On this build we start running 8 jobs (conan-package-tools). If you search for package id *7997b2d55906abb9d381551aae242b4fb12b16eb* you will be able to see 2 packages using same id. The first package uses boost_*:shared=True and the second uses boost_*:shared=False

**Comment by memsharded** [comment]
HI @uilianries could you please verify the header_only value for those packages? Also, the full settings, or better the conaninfo.txt generated could be valuable. It could be something of the logic of the ``package_id()``, but it is indirected to the "optional_package_id()", in which package it is implemented?

**Comment by uilianries** [comment]
Before to read your comment I collected some new information.

Running boost_level8group/1.66.0@bincrafters/testing on lasote/conangcc6
```
conan create . bincrafters/testing -o boost_*:shared=False -s build_type=Debug -s arch=x86_64
self.options.values.sha: dec258f52d7fbdc727ec855c4f01c6ea1fbebd96
self.settings.values.sha: a8d08f4bdb8a8acb1c6d222aec1e0a22832c7fcc
package_id: 58e66c5755b9eb1f3b6e68b70ed92e6ac5c1684a

conan create . bincrafters/testing -o boost_*:shared=False -s build_type=Debug -s arch=x86
self.options.values.sha: dec258f52d7fbdc727ec855c4f01c6ea1fbebd96
self.settings.values.sha: f15349e9dcb23b1fb9ff0194a644656b09dec080
package_id: bf58d220ad869d99debf26b4229194054ba7a256

conan create . bincrafters/testing -o boost_*:shared=True -s build_type=Debug -s arch=x86
self.options.values.sha: 014a8596924f47db5825e2b38c39f4fc2daa5622
self.settings.values.sha: f15349e9dcb23b1fb9ff0194a644656b09dec080
package_id: bf58d220ad869d99debf26b4229194054ba7a256

conan create . bincrafters/testing -o boost_*:shared=True -s build_type=Debug -s arch=x86_64
self.options.values.sha: 014a8596924f47db5825e2b38c39f4fc2daa5622
self.settings.values.sha: a8d08f4bdb8a8acb1c6d222aec1e0a22832c7fcc
package_id: 58e66c5755b9eb1f3b6e68b70ed92e6ac5c1684a

conan create . bincrafters/testing -o boost_*:shared=True -s build_type=Release -s arch=x86_64
self.options.values.sha: 014a8596924f47db5825e2b38c39f4fc2daa5622
self.settings.values.sha: fedb4c9d2e5b701bca67c2028747f524852d272b
package_id: dc6456f4704958f26592d9cfc6fa47e91447bd41

conan create . bincrafters/testing -o boost_*:shared=True -s build_type=Release -s arch=x86
self.options.values.sha: 014a8596924f47db5825e2b38c39f4fc2daa5622
self.settings.values.sha: 16d0f8c0499a6f4337674481cad90a9a7a6d5835
package_id: 9f4335386f678dc053976aeaed83b4a6e4bf4b5e

conan create . bincrafters/testing -o boost_*:shared=False -s build_type=Release -s arch=x86
self.options.values.sha: dec258f52d7fbdc727ec855c4f01c6ea1fbebd96
self.settings.values.sha: 16d0f8c0499a6f4337674481cad90a9a7a6d5835
package_id: 9f4335386f678dc053976aeaed83b4a6e4bf4b5e

conan create . bincrafters/testing -o boost_*:shared=False -s build_type=Release -s arch=x86_64
self.options.values.sha: dec258f52d7fbdc727ec855c4f01c6ea1fbebd96
self.settings.values.sha: fedb4c9d2e5b701bca67c2028747f524852d272b
package_id: dc6456f4704958f26592d9cfc6fa47e91447bd41
```
package_id was collected by `self.info.package_id()
self.options.values.sha was collected by `self.options.values.sha`
self.settings.values.sha was collected by `self.settings.values.sha`

Now I chose two results to compare:
```
conan create . bincrafters/testing -o boost_*:shared=False -s build_type=Debug -s arch=x86_64
self.options.values.sha: dec258f52d7fbdc727ec855c4f01c6ea1fbebd96
self.settings.values.sha: a8d08f4bdb8a8acb1c6d222aec1e0a22832c7fcc
package_id: 58e66c5755b9eb1f3b6e68b70ed92e6ac5c1684a

conan create . bincrafters/testing -o boost_*:shared=True -s build_type=Debug -s arch=x86_64
self.options.values.sha: 014a8596924f47db5825e2b38c39f4fc2daa5622
self.settings.values.sha: a8d08f4bdb8a8acb1c6d222aec1e0a22832c7fcc
package_id: 58e66c5755b9eb1f3b6e68b70ed92e6ac5c1684a
```
Both uses the same settings, but the options are not the same. The package id is the same for both two. Looks like when options contains `*` the package id skip it on append.

I used the latest Conan version (1.0.3)

**Comment by solvingj** [comment]
After staring at it for a bit of time, I think @wpalfi's findings explain our issue.  Our `conan-package-tools` template says `shared_option_name="boost_*:shared"`.  If `shared` isn't getting applied to the current package, then all the builds which `conan-package-tools` is launching with `boost_*:shared=True` are actually being built with `shared=False` and thus, each variant gets built with `shared=False` twice.  This would explain how we end up with only 4 binaries instead of 8, and the matching hashes.

**Comment by wpalfi** [comment]
@bincrafters You may also be very interested in #2393 
I actually stumbled upon #2391 and #2393 when I tried to build bincrafters boost packages with *:shared=True.

**Comment by memsharded** [comment]
Ok, got it. Will give it top priority and try to fix it for 1.0.4 on  Monday.
Thanks all for the feedback!

**Comment by solvingj** [comment]
FYI, we've worked around it.  Our shared code for building the boost packages with `conan-package-tools` now passes both shared options when building.  Whenever it gets patched in conan, great... but it's not blocking us at the moment.
